### PR TITLE
[IMP] Project: Generic improvements

### PR DESCRIPTION
**Pull request by hugocarlier**
- product form view: set 'hours' as the UoM if the service is either
prepaid or invoice based on timesheets and if sale_timesheet is
installed.

- Details: Before this commit, the default UoM was the same for every
invoice policy or product type selected. With this commit the UoM is
automatically changed to 'Hours' when the product is of type service
and the invoice policy is either prepaid (A) or based on
timesheet (B). The value of the UoM can still be changed manually. The
following corner cases are managed:
    - The user edits (or creates) a product, changes its type to
    service then to storable/consumable without saving it. The UoM
    should then be set to its initial value (before change) or its
    default value.
    - The user edits (or creates) a product, changes its invoice
    policy to A or B then again to another one without saving it. The
    UoM should then be set to its initial value (before change) or its
    default value.
    - The user edits a product of type service with invoice policy set
    to A (or B) and UoM that has manually been changed to something
    else than 'Hours'. He then modifies the invoice policy to B (or A)
    and then again to A (or B) without saving the changes. The UoM should
    then be set back to its initial value.

Task-3072818

### CODE REVIEW

File: addons/project/models/project.py [hunk 1]
```diff
@@ -2125,6 +2138,37 @@ def _unlink_except_recurring(self):
...
+                    converted_dom = dom[2] if isinstance(dom[2], set) or isinstance(dom[2], list) or isinstance(dom[2], tuple) else json.loads(dom[2]) #converting string representation of a list to a list
+                    if isinstance(converted_dom, list) or isinstance(converted_dom, tuple) or isinstance(converted_dom, set) and len(converted_dom) >= 2:
+                        return users
+                    elif isinstance(converted_dom, list) or isinstance(converted_dom, tuple) or isinstance(converted_dom, set) and len(converted_dom) == 1:
+                        if display_project_id:
+                            return users
+                        else:
+                            display_project_id = converted_dom[0]
+        if not display_project_id:
+            return users
```
Comment (auon-odoo) [inline addons/project/models/project.py:2165]: Not again :sweat_smile: 
But I think we can just use the `default_project_id` key of the context. Not sure really.

File: addons/project/models/project.py [hunk 2]
```diff
@@ -1676,6 +1678,17 @@ def _compute_stage_display(self):
         for task in self:
             task.stage_display = task.stage_id.name if task.project_id else task.personal_stage_id.stage_id.name
 
+    @api.depends('depend_on_ids', 'is_blocked')
+    def _compute_blocking_task_display(self):
+        for record in self:
+            if not record.is_blocked:
```
Comment (auon-odoo) [inline addons/project/models/project.py:1684]:
```suggestion
            if not record.is_blocked or task.is_closed:
```
Because we won't use it if it is closed, will we?

File: addons/project/models/project.py [hunk 3]
```diff
@@ -1676,6 +1678,17 @@ def _compute_stage_display(self):
...
+                record.blocking_task_display = ""
+            else:
+                blocking_tasks = [task for task in record.depend_on_ids if not task.is_closed]
```
Comment (auon-odoo) [inline addons/project/models/project.py:1687]:
```suggestion
                blocking_tasks = record.depend_on_ids.filtered(lambda task: not task.is_closed)
```

File: addons/project/models/project.py [hunk 4]
```diff
@@ -1676,6 +1678,17 @@ def _compute_stage_display(self):
...
+                record.blocking_task_display = "Task " + blocking_tasks[0].name
+                record.blocking_task_display += " (and " + str(len(blocking_tasks)-1) + " other tasks)" if len(blocking_tasks) > 2 else ""
+                record.blocking_task_display += " (and one other task)" if len(blocking_tasks) == 2 else ""
```
Comment (auon-odoo) [inline addons/project/models/project.py:1690]:
```suggestion
                record.blocking_task_display += " (and " + str(len(blocking_tasks)-1) + " other tasks)" if len(blocking_tasks) > 1 else ""
```
I think we should keep it simple. Also, will that be translated?

File: addons/project/views/project_views.xml [hunk 5]
```diff
@@ -1098,6 +1099,9 @@
                         <p>Edit recurring task</p>
                         <field name="recurrence_update" widget="radio"/>
                     </div>
+                    <div class="alert alert-warning oe_edit_only mb-0" role="status" attrs="{'invisible': ['|', ('is_blocked', '=', False), ('is_closed', '=', True)]}" groups="project.group_project_task_dependencies">
+                        <p><b><field name="blocking_task_display"/></b> should be completed before starting working on this task.</p>
```
Comment (auon-odoo) [inline addons/project/views/project_views.xml:1103]: Shouldn't we put this rest of the sentence in the compute?

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you :zap:

File: addons/project/models/project.py (hunk 1 above)
Comment (huca-odoo) [reply addons/project/models/project.py]: The logic I used was not entirely adequate. I've rewritten it in a (I think) better way. To be short:
1- I first check the key default_project_id (set if I'm in the view of a project).
2- Then I check the domain is the key is not set to see tasks of which project are displayed.

File: addons/project/models/project.py (hunk 2 above)
Comment (huca-odoo) [reply addons/project/models/project.py:1684]: I actually have a check on the is_closed in the xml but I guess doing this spare some calculation.

File: addons/project/models/project.py (hunk 4 above)
Comment (huca-odoo) [reply addons/project/models/project.py]: You're right, I wasn't taking translation into account. I changed this method following the guidelines that I found here: https://www.odoo.com/documentation/16.0/developer/howtos/translations.html

**Review by huca-odoo — COMMENTED** [review]
Thank you for your comments, I've updated my code :)

File: addons/sale_timesheet/models/product.py [hunk 6]
```diff
@@ -68,6 +68,17 @@ def _compute_product_tooltip(self):
                         "project for the order to track the time spent."
                     )
 
+    @api.onchange('type', 'service_type', 'service_policy')
+    def _update_uom_id(self):
+        for record in self:
+            if record.type == 'service' and record.service_type == 'timesheet' and \
+               (not record._origin or not record._origin.service_policy or record.service_policy != record._origin.service_policy):
```
Comment (auon-odoo) [inline addons/sale_timesheet/models/product.py:75]:
```suggestion
            if record.type == 'service' and record.service_type == 'timesheet' and \
               not (record._origin.service_policy and record.service_policy == record._origin.service_policy):
```
- Same comment as below
- "Mise en évidence" du not

**Review by auon-odoo — CHANGES_REQUESTED** [review]
Thank you! :zap: And sorry for the delay

File: addons/sale_timesheet/models/product.py [hunk 7]
```diff
@@ -68,6 +68,17 @@ def _compute_product_tooltip(self):
...
+                record.uom_id = self.env.ref('uom.product_uom_hour')
+            elif record._origin and record._origin.uom_id:
```
[the remark this answers is not in the document]
Comment (huca-odoo) [reply addons/sale_timesheet/models/product.py:77]: I did that to avoid an error when accessing _origin.uom_id if _origin does not exist but I guess it's not necessary.

File: addons/sale_timesheet/models/product.py (hunk 6 above)
Comment (huca-odoo) [reply addons/sale_timesheet/models/product.py:75]: I'll update that ;)

**Review by huca-odoo — COMMENTED** [review]
Thank you for your feedback! I've added some replies to your comments. I will see with Alexandra for the two tasks conflicting and take some time to re-create the branch because the rebase is problematic...

File: addons/project/models/project.py [hunk 8]
```diff
@@ -1684,6 +1686,24 @@ def _compute_stage_display(self):
...
+                if nb_blocking_tasks == 1:
+                    record.blocking_task_display = _("Task <b>%s</b> should be completed before starting working on this task.",\
+                                                     blocking_tasks[0].name)
+                elif nb_blocking_tasks == 2:
+                    record.blocking_task_display = _("Task <b>%s</b> (and one other task) should be completed before starting working on this task.",\
+                                                     blocking_tasks[0].name)
+                elif nb_blocking_tasks > 2:
+                    record.blocking_task_display = _("Task <b>%s</b> (and %d other tasks) should be completed before starting working on this task.",\
+                                                     blocking_tasks[0].name, nb_blocking_tasks-1)
+
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/project/models/project.py:1706]: Indeed, the idea is to keep it simple.
In this particular case, it seems to me that both "sentences" can be translated separately without trouble :monocle_face:

**Comment by huca-odoo** [comment]
@auon-odoo : After having checked with Alexandra, we will drop first and second commits that have already be addressed by other tasks. I updated the third one based on your comments. Thank you for your feedback!

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left one comment. 🙂

File: addons/sale_timesheet/models/product.py [hunk 9]
```diff
@@ -68,6 +68,17 @@ def _compute_product_tooltip(self):
...
+    @api.onchange('type', 'service_type', 'service_policy')
+    def _update_uom_id(self):
+        for record in self:
+            if record.type == 'service' and record.service_type == 'timesheet' and \
+               not (record._origin.service_policy and record.service_policy == record._origin.service_policy):
+                record.uom_id = self.env.ref('uom.product_uom_hour')
+            elif record._origin.uom_id:
+                record.uom_id = record._origin.uom_id
+            else:
+                record.uom_id = self._get_default_uom_id()
```
[the remark this answers is not in the document]
Comment (huca-odoo) [reply addons/sale_timesheet/models/product.py:80]: > Why did you choose to do an onchange instead of a compute method? monocle_face

@xavierbol Thank you again for the review :pray: To get back on this: I chose to use an onchange method to avoid undesired behaviors that could have happened imo with a compute method with a depends decorator.

The way I understand the spec, this change is more about suggesting a default time uom to the user when he selects a service product, rather than setting this blindly in every circumstances. For example, if the service type of a record is changed anywhere using the orm, I think it is preferable to not change the uom than setting it arbitrary to hour, without even changing the value of the price of the product.

In summary, I don't really see a use case where we would prefer a compute method to this onchange implementation. Let me know your opinion about this.

[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/sale_timesheet/models/product.py:80]: Indeed, we discussed about it and indeed we will keep the onchange, however, we have to call it `_onchange_uom_id` (maybe that onchange already exists in sale module)

Also, the onchange has to be defined in `product.template` as well, to be able to have the same behavior in that model since `product.product` is a `product.template` with additional fields and features (thanks to `inherits`).

[the remark this answers is not in the document]
Comment (huca-odoo) [reply addons/sale_timesheet/models/product.py]: > Indeed, we discussed about it and indeed we will keep the onchange, however, we have to call it `_onchange_uom_id` (maybe that onchange already exists in sale module)
> 
> Also, the onchange has to be defined in `product.template` as well, to be able to have the same behavior in that model since `product.product` is a `product.template` with additional fields and features (thanks to `inherits`).

I've changed the onchange method as discussed and added some tests. Let me know if you have other comments.

**Review by xavierbol — APPROVED** [review]
LGTM, thanks for your work. 🙂 

robodoo r+
